File: 2.0/WearOS/WatchSprings-2.0/app/src/main/java/copyright/markmorris/allrightsreserved/watchsprings/api/FortniteApi.kt
// WatchSprings 2.0 © 2025 Mark S. Morris. All rights reserved.
//
// https://www.watchsprings.com
//
// WatchSprings is Fan Content pursuant to the Epic Games Fan Content Policy. Portions of the materials used are trademarks and/or copyrighted works of Epic Games, Inc.

package copyright.markmorris.allrightsreserved.watchsprings

//Get API key at https://fortniteapi.io/

import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.engine.cio.CIO
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.request.get
import io.ktor.client.request.header
import io.ktor.client.request.parameter
import io.ktor.serialization.kotlinx.json.json
import kotlinx.serialization.json.Json

class FortniteApiClient()
{
    private var apiKey =  "Get API key at https://fortniteapi.io/"

    private val client = HttpClient(CIO) {
        install(ContentNegotiation) {
            json(Json { ignoreUnknownKeys = true
                        coerceInputValues = true })     // turns null into ""
        }
    }

    private val baseUrl = "https://fortniteapi.io"

    suspend fun getBattleRoyaleNews(): NewsResponse =
        client.get("$baseUrl/v1/news") {
            parameter("lang", "en")
            parameter("type", "br")
            header("Authorization", apiKey)
        }.body()

    suspend fun upcomingItems(): ItemsResponse =
        client.get("$baseUrl/v2/items/upcoming") {
            parameter("lang", "en")
            header("Authorization", apiKey)
        }.body()

    suspend fun getDailyShop(): ShopResponse =
        client.get("$baseUrl/v2/shop") {
            parameter("lang", "en")
            header("Authorization", apiKey)
        }.body()

    suspend fun getBattlePassRewards(): BattlePassResponse =
        client.get("$baseUrl/v2/battlepass") {
            parameter("lang", "en")
            parameter("season", "current")
            header("Authorization", apiKey)
        }.body()

    fun getMapImageUrl(showPOI: Boolean = false): String =
        "https://media.fortniteapi.io/images/map.png?showPOI=$showPOI"

    suspend fun getServerStatus(): StatusResponse =
        client.get("https://status.epicgames.com/api/v2/summary.json")
            .body()

}

File: 2.0/watchOS/WatchSprings-2.0/WatchSprings Watch App/api/FortniteAPI.swift
// WatchSprings 2.0 © 2025 Mark S. Morris. All rights reserved.
//
// https://www.watchsprings.com
//
// WatchSprings is Fan Content pursuant to the Epic Games Fan Content Policy. Portions of the materials used are trademarks and/or copyrighted works of Epic Games, Inc.

import Foundation


class FortniteApiClient
{
    static let instance = FortniteApiClient()
    
    private var apiKey = "Get API key at https://fortniteapi.io"
    
    private let baseUrl = "https://fortniteapi.io"
    
    private let session = URLSession.shared
    private let jsonDecoder: JSONDecoder
    
    init()
    {
        // Setup JSON decoder
        jsonDecoder = JSONDecoder()
        jsonDecoder.keyDecodingStrategy = .useDefaultKeys
    }
    
    // Helper method to create URLRequest with common headers
    private func createRequest(for endpoint: String, queryItems: [URLQueryItem]) -> URLRequest
    {
        var components = URLComponents(string: baseUrl + endpoint)!
        components.queryItems = queryItems
        
        var request = URLRequest(url: components.url!)
        request.addValue(apiKey, forHTTPHeaderField: "Authorization")
        return request
    }
    
    // Generic fetch method
    private func fetch<T: Decodable>(request: URLRequest) async throws -> T {
        let (data, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw URLError(.badServerResponse)
        }
        
        return try jsonDecoder.decode(T.self, from: data)
    }
    
    // API Methods
    func getBattleRoyaleNews() async throws -> NewsResponse {
        let queryItems = [URLQueryItem(name: "lang", value: "en"),
                          URLQueryItem(name: "type", value: "br")]
        let request = createRequest(for: "/v1/news", queryItems: queryItems)
        return try await fetch(request: request)
    }
    
    func upcomingItems() async throws -> ItemsResponse {
        let queryItems = [URLQueryItem(name: "lang", value: "en")]
        let request = createRequest(for: "/v2/items/upcoming", queryItems: queryItems)
        return try await fetch(request: request)
    }
    
    func getDailyShop() async throws -> ShopResponse {
        let queryItems = [URLQueryItem(name: "lang", value: "en")]
        let request = createRequest(for: "/v2/shop", queryItems: queryItems)
        return try await fetch(request: request)
    }
    
    func getBattlePassRewards() async throws -> BattlePassResponse {
        let queryItems = [URLQueryItem(name: "lang", value: "en"),
                          URLQueryItem(name: "season", value: "current")]
        let request = createRequest(for: "/v2/battlepass", queryItems: queryItems)
        return try await fetch(request: request)
    }
    
    func getMapImageUrl(showPOI: Bool = false) -> URL?
    {
        return URL(string: "https://media.fortniteapi.io/images/map.png?showPOI=\(showPOI)")
    }
    
    func getServerStatus() async throws -> StatusResponse {
        let statusUrl = URL(string: "https://status.epicgames.com/api/v2/summary.json")!
        let request = URLRequest(url: statusUrl)
        return try await fetch(request: request)
    }
}

